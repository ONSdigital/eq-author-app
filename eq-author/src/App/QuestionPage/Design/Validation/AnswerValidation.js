import React from "react";
import { kebabCase, get, startCase } from "lodash";
import styled from "styled-components";
import CustomPropTypes from "custom-prop-types";
import PropTypes from "prop-types";
import { connect } from "react-redux";

import { gotoTab } from "redux/tabs/actions";

import SidebarButton, { Title, Detail } from "components/Buttons/SidebarButton";
import ModalWithNav from "components/Modals/ModalWithNav";

import MinValueValidation from "App/QuestionPage/Design/Validation/MinValue";
import MaxValueValidation from "App/QuestionPage/Design/Validation/MaxValue";
import {
  EarliestDate,
  LatestDate,
  MinDuration,
  MaxDuration
} from "App/QuestionPage/Design/Validation/Date";

import ValidationContext from "App/QuestionPage/Design/Validation/ValidationContext";
import DatePreview from "App/QuestionPage/Design/Validation/Date/DatePreview";
import DurationPreview from "App/QuestionPage/Design/Validation/Date/DurationPreview";

import { CURRENCY, DATE, DATE_RANGE, NUMBER } from "constants/answer-types";
import { colors } from "constants/theme";

const Container = styled.div`
  margin-top: 1em;
  border-top: 1px solid ${colors.lightGrey};
  padding: 1em 0;
`;

const validationTypes = [
  {
    id: "minValue",
    title: "Min Value",
    render: () => <MinValueValidation />,
    types: [CURRENCY, NUMBER],
    preview: ({ custom }) => custom
  },
  {
    id: "maxValue",
    title: "Max Value",
    render: () => <MaxValueValidation />,
    types: [CURRENCY, NUMBER],
    preview: ({ custom, previousAnswer }) =>
      custom ? custom : get(previousAnswer, "displayName")
  },
  {
    id: "earliestDate",
    title: "Earliest Date",
    render: () => <EarliestDate />,
    types: [DATE, DATE_RANGE],
    preview: DatePreview
  },
  {
    id: "latestDate",
    title: "Latest Date",
    render: () => <LatestDate />,
    types: [DATE, DATE_RANGE],
    preview: DatePreview
  },
  {
    id: "minDuration",
    title: "Min Duration",
    render: () => <MinDuration />,
    types: [DATE_RANGE],
    preview: DurationPreview
  },
  {
    id: "maxDuration",
    title: "Max Duration",
    render: () => <MaxDuration />,
    types: [DATE_RANGE],
    preview: DurationPreview
  }
];

const getValidationsForType = type =>
  validationTypes.filter(({ types }) => types.includes(type));

const validations = [NUMBER, CURRENCY, DATE, DATE_RANGE].reduce(
  (hash, type) => ({
    ...hash,
    [type]: getValidationsForType(type)
  }),
  {}
);

export class UnconnectedAnswerValidation extends React.Component {
  state = {
    modalIsOpen: false,
    answerId: null
  };

  constructor(props) {
    super(props);
    this.modalId = `modal-validation-${props.answer.id}`;
  }

  handleModalClose = () => this.setState({ modalIsOpen: false });

  renderButton = ({ id, title, value, enabled }) => (
    <SidebarButton
      key={id}
      data-test={`sidebar-button-${kebabCase(title)}`}
      onClick={() => {
        this.props.gotoTab(this.modalId, id);
        this.setState({ modalIsOpen: true });
      }}
    >
      <Title>{title}</Title>
      {enabled && value && <Detail>{value}</Detail>}
    </SidebarButton>
  );

  render() {
    const { answer } = this.props;
    const validValidationTypes = validations[answer.type] || [];
    if (validValidationTypes.length === 0) {
      return null;
    }

    return (
      <Container>
        <ValidationContext.Provider value={{ answer }}>
          {validValidationTypes.map(validationType => {
            const validation = get(
              answer,
              `validation.${validationType.id}`,
              {}
            );
            const { enabled, previousAnswer, metadata } = validation;
            const value = enabled ? validationType.preview(validation) : "";

            return this.renderButton({
              ...validationType,
              value,
              enabled,
              previousAnswer,
              metadata
            });
          })}
          <ModalWithNav
            id={this.modalId}
            onClose={this.handleModalClose}
            navItems={validValidationTypes}
            title={`${startCase(answer.type)} validation`}
            isOpen={this.state.modalIsOpen}
          />
        </ValidationContext.Provider>
      </Container>
    );
  }
}

UnconnectedAnswerValidation.propTypes = {
  answer: CustomPropTypes.answer,
  gotoTab: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  tabsState: state.tabs
});

export default connect(
  mapStateToProps,
  { gotoTab }
)(UnconnectedAnswerValidation);
