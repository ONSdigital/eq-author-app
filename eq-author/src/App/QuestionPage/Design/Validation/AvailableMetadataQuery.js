import React from "react";
import PropTypes from "prop-types";
import { Query } from "react-apollo";
import gql from "graphql-tag";

import AvailableMetadata from "graphql/fragments/available-metadata.graphql";

const GET_AVAILABLE_METADATA = gql`
  query GetAvailableMetadata($id: ID!) {
    answer(id: $id) {
      id
      displayName
      ... on BasicAnswer {
        validation {
          ... on DateValidation {
            earliestDate {
              id
              availableMetadata {
                ...AvailableMetadata
              }
            }
            latestDate {
              id
              availableMetadata {
                ...AvailableMetadata
              }
            }
          }
        }
      }
      ... on CompositeAnswer {
        validation {
          ... on DateRangeValidation {
            earliestDate {
              id
              availableMetadata {
                ...AvailableMetadata
              }
            }
            latestDate {
              id
              availableMetadata {
                ...AvailableMetadata
              }
            }
          }
        }
      }
    }
  }
  ${AvailableMetadata}
`;

const AvailableMetadataQuery = ({ answerId, children }) => (
  <Query
    query={GET_AVAILABLE_METADATA}
    variables={{ id: answerId }}
    fetchPolicy="cache-and-network"
  >
    {children}
  </Query>
);

AvailableMetadataQuery.propTypes = {
  answerId: PropTypes.string.isRequired,
  children: PropTypes.func.isRequired
};

export default AvailableMetadataQuery;
