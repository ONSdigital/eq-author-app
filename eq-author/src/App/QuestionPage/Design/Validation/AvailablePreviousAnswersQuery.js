import React from "react";
import PropTypes from "prop-types";
import { Query } from "react-apollo";
import gql from "graphql-tag";

import AvailableAnswers from "graphql/fragments/available-answers.graphql";

const GET_AVAILABLE_PREVIOUS_ANSWERS = gql`
  query GetAvailablePreviousAnswers($id: ID!) {
    answer(id: $id) {
      id
      displayName
      ... on BasicAnswer {
        validation {
          ... on NumberValidation {
            maxValue {
              id
              availablePreviousAnswers {
                ...AvailableAnswers
              }
            }
          }
          ... on DateValidation {
            earliestDate {
              id
              availablePreviousAnswers {
                ...AvailableAnswers
              }
            }
            latestDate {
              id
              availablePreviousAnswers {
                ...AvailableAnswers
              }
            }
          }
        }
      }
    }
  }
  ${AvailableAnswers}
`;

const AvailablePreviousAnswersQuery = ({ answerId, children }) => (
  <Query
    query={GET_AVAILABLE_PREVIOUS_ANSWERS}
    variables={{ id: answerId }}
    fetchPolicy="cache-and-network"
  >
    {children}
  </Query>
);

AvailablePreviousAnswersQuery.propTypes = {
  answerId: PropTypes.string.isRequired,
  children: PropTypes.func.isRequired
};

export default AvailablePreviousAnswersQuery;
