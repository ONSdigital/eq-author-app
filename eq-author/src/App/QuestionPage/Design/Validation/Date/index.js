import builder from "App/QuestionPage/Design/Validation/Date/builder";
import {
  dateReadToWriteMapper,
  durationReadToWriteMapper
} from "App/QuestionPage/Design/Validation/Date/readToWriteMapper";
import DateValidation from "App/QuestionPage/Design/Validation/Date/DateValidation";
import DurationValidation from "App/QuestionPage/Design/Validation/Date/DurationValidation";

import EarliestDateValidationRule from "graphql/fragments/earliest-date-validation-rule.graphql";
import LatestDateValidationRule from "graphql/fragments/latest-date-validation-rule.graphql";
import MinDurationValidationRule from "graphql/fragments/min-duration-validation-rule.graphql";
import MaxDurationValidationRule from "graphql/fragments/max-duration-validation-rule.graphql";

export const LatestDate = builder(
  "Latest date",
  "latest-date-validation",
  "latestDate",
  "latestDateInput",
  LatestDateValidationRule,
  dateReadToWriteMapper,
  "date"
)(DateValidation);

export const EarliestDate = builder(
  "Earliest date",
  "earliest-date-validation",
  "earliestDate",
  "earliestDateInput",
  EarliestDateValidationRule,
  dateReadToWriteMapper,
  "date"
)(DateValidation);

export const MinDuration = builder(
  "Min duration",
  "min-duration-validation",
  "minDuration",
  "minDurationInput",
  MinDurationValidationRule,
  durationReadToWriteMapper,
  "duration"
)(DurationValidation);

export const MaxDuration = builder(
  "Max duration",
  "max-duration-validation",
  "maxDuration",
  "maxDurationInput",
  MaxDurationValidationRule,
  durationReadToWriteMapper,
  "duration"
)(DurationValidation);
