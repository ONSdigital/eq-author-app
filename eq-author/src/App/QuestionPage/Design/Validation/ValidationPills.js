import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { compact, isFunction } from "lodash";

import PillTabs from "components/PillTabs";

import * as entityTypes from "constants/validation-entity-types";

export const Pills = styled(PillTabs)`
  width: 5em;
`;

const PillTabContent = styled.div`
  margin-top: 3em;
  margin-bottom: 0;
  width: 100%;
`;

export const ValidationPills = ({
  entityType,
  onEntityTypeChange,
  PreviousAnswer,
  Metadata,
  Now,
  Custom
}) => (
  <Pills
    value={entityType}
    onChange={onEntityTypeChange}
    options={compact([
      isFunction(Now)
        ? {
            id: "Now",
            title: "Start date",
            render: () => (
              <PillTabContent>
                <Now />
              </PillTabContent>
            )
          }
        : null,
      isFunction(PreviousAnswer)
        ? {
            id: "PreviousAnswer",
            title: "Previous answer",
            render: () => (
              <PillTabContent>
                <PreviousAnswer />
              </PillTabContent>
            )
          }
        : null,
      isFunction(Metadata)
        ? {
            id: "Metadata",
            title: "Metadata",
            render: () => (
              <PillTabContent>
                <Metadata />
              </PillTabContent>
            )
          }
        : null,
      isFunction(Custom)
        ? {
            id: "Custom",
            title: "Custom",
            render: () => (
              <PillTabContent>
                <Custom />
              </PillTabContent>
            )
          }
        : null
    ])}
  />
);

ValidationPills.propTypes = {
  entityType: PropTypes.oneOf(Object.values(entityTypes)).isRequired,
  onEntityTypeChange: PropTypes.func.isRequired,
  PreviousAnswer: PropTypes.func,
  Metadata: PropTypes.func,
  Custom: PropTypes.func,
  Now: PropTypes.func
};
