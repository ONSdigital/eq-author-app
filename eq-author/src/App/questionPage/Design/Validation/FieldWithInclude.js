import React, { Fragment } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { Label, Field } from "components/Forms";

import VisuallyHidden from "components/VisuallyHidden";
import ToggleSwitch from "components/buttons/ToggleSwitch";

const InlineField = styled(Field)`
  display: flex;
  flex-direction: row;
  align-items: center;
  margin-left: -0.8em;
  padding: 0 0.5em;
`;

const FieldWithInclude = ({ children, id, name, onChange, ...otherProps }) => (
  <Fragment>
    <VisuallyHidden>
      <Label hidden>Value</Label>
    </VisuallyHidden>
    {children}
    <InlineField>
      <ToggleSwitch id={id} name={name} onChange={onChange} {...otherProps} />
      <Label inline htmlFor={id}>
        Include this number
      </Label>
    </InlineField>
  </Fragment>
);

FieldWithInclude.propTypes = {
  children: PropTypes.node.isRequired,
  id: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  onChange: PropTypes.func.isRequired
};

export default FieldWithInclude;
