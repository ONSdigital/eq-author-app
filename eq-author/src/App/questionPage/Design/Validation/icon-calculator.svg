<?xml version="1.0" encoding="UTF-8"?>
<svg width="32px" height="32px" viewBox="0 0 32 32" version="1.1" xmlns="http://www.w3.org/2000/svg" xmlns:xlink="http://www.w3.org/1999/xlink">
    <!-- Generator: Sketch 52.6 (67491) - http://www.bohemiancoding.com/sketch -->
    <title>Artboard</title>
    <desc>Created with Sketch.</desc>
    <g id="Artboard" stroke="none" stroke-width="1" fill="none" fill-rule="evenodd">
        <path d="M12.9176353,23.7059186 C12.9176353,22.8510433 12.2313269,22.1647349 11.3764516,22.1647349 C10.5215762,22.1647349 9.83526785,22.8510433 9.83526785,23.7059186 C9.83526785,24.560794 10.5215762,25.2471024 11.3764516,25.2471024 C12.2313269,25.2471024 12.9176353,24.560794 12.9176353,23.7059186 Z M17.5411865,23.7059186 C17.5411865,22.8510433 16.8548781,22.1647349 16.0000028,22.1647349 C15.1451274,22.1647349 14.458819,22.8510433 14.458819,23.7059186 C14.458819,24.560794 15.1451274,25.2471024 16.0000028,25.2471024 C16.8548781,25.2471024 17.5411865,24.560794 17.5411865,23.7059186 Z M12.9176353,19.0823675 C12.9176353,18.2274921 12.2313269,17.5411837 11.3764516,17.5411837 C10.5215762,17.5411837 9.83526785,18.2274921 9.83526785,19.0823675 C9.83526785,19.9372428 10.5215762,20.6235512 11.3764516,20.6235512 C12.2313269,20.6235512 12.9176353,19.9372428 12.9176353,19.0823675 Z M17.5411865,19.0823675 C17.5411865,18.2274921 16.8548781,17.5411837 16.0000028,17.5411837 C15.1451274,17.5411837 14.458819,18.2274921 14.458819,19.0823675 C14.458819,19.9372428 15.1451274,20.6235512 16.0000028,20.6235512 C16.8548781,20.6235512 17.5411865,19.9372428 17.5411865,19.0823675 Z M12.9176353,14.4588163 C12.9176353,13.6039409 12.2313269,12.9176325 11.3764516,12.9176325 C10.5215762,12.9176325 9.83526785,13.6039409 9.83526785,14.4588163 C9.83526785,15.3136916 10.5215762,16 11.3764516,16 C12.2313269,16 12.9176353,15.3136916 12.9176353,14.4588163 Z M22.1647377,23.7059186 L22.1647377,19.0823675 C22.1647377,18.2395326 21.4663888,17.5411837 20.6235539,17.5411837 C19.7807191,17.5411837 19.0823702,18.2395326 19.0823702,19.0823675 L19.0823702,23.7059186 C19.0823702,24.5487535 19.7807191,25.2471024 20.6235539,25.2471024 C21.4663888,25.2471024 22.1647377,24.5487535 22.1647377,23.7059186 Z M17.5411865,14.4588163 C17.5411865,13.6039409 16.8548781,12.9176325 16.0000028,12.9176325 C15.1451274,12.9176325 14.458819,13.6039409 14.458819,14.4588163 C14.458819,15.3136916 15.1451274,16 16.0000028,16 C16.8548781,16 17.5411865,15.3136916 17.5411865,14.4588163 Z M22.1647377,10.605857 L22.1647377,7.52348951 C22.1647377,7.10207209 21.9107926,6.75289765 21.6043072,6.75289765 L10.3956983,6.75289765 C10.0892129,6.75289765 9.83526785,7.10207209 9.83526785,7.52348951 L9.83526785,10.605857 C9.83526785,11.0272744 10.0892129,11.3764488 10.3956983,11.3764488 L21.6043072,11.3764488 C21.9107926,11.3764488 22.1647377,11.0272744 22.1647377,10.605857 Z M22.1647377,14.4588163 C22.1647377,13.6039409 21.4784293,12.9176325 20.6235539,12.9176325 C19.7686786,12.9176325 19.0823702,13.6039409 19.0823702,14.4588163 C19.0823702,15.3136916 19.7686786,16 20.6235539,16 C21.4784293,16 22.1647377,15.3136916 22.1647377,14.4588163 Z M23.7059214,6.75289765 L23.7059214,25.2471024 C23.7059214,26.0899372 23.1687297,26.7882861 22.520395,26.7882861 L9.47960498,26.7882861 C8.83127025,26.7882861 8.29407862,26.0899372 8.29407862,25.2471024 L8.29407862,6.75289765 C8.29407862,5.9100628 8.83127025,5.21171392 9.47960498,5.21171392 L22.520395,5.21171392 C23.1687297,5.21171392 23.7059214,5.9100628 23.7059214,6.75289765 Z" id="ï‡¬" fill="#666666"></path>
    </g>
</svg>