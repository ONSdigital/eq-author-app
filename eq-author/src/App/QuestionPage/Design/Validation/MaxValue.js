import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import { withApollo } from "react-apollo";
import { propType } from "graphql-anywhere";
import { flowRight, get, inRange, isNaN } from "lodash";

import { Field, Label } from "components/Forms";
import { Grid, Column } from "components/Grid";
import ToggleSwitch from "components/Buttons/ToggleSwitch";

import PreviousAnswerContentPicker from "App/QuestionPage/Design/Validation/PreviousAnswerContentPicker";
import DisabledMessage from "App/QuestionPage/Design/Validation/DisabledMessage";
import { ValidationPills } from "App/QuestionPage/Design/Validation/ValidationPills";
import ValidationTitle from "App/QuestionPage/Design/Validation/ValidationTitle";
import ValidationView from "App/QuestionPage/Design/Validation/ValidationView";
import ValidationContext from "App/QuestionPage/Design/Validation/ValidationContext";
import ValidationInput from "App/QuestionPage/Design/Validation/ValidationInput";
import PathEnd from "App/QuestionPage/Design/Validation/path-end.svg?inline";

import withUpdateAnswerValidation from "App/QuestionPage/Design/Validation/withUpdateAnswerValidation";
import withToggleAnswerValidation from "App/QuestionPage/Design/Validation/withToggleAnswerValidation";

import MaxValueValidationRule from "graphql/fragments/max-value-validation-rule.graphql";

import * as answerTypes from "constants/answer-types";

const InlineField = styled(Field)`
  display: flex;
  flex-direction: row;
  align-items: center;
  margin-left: -0.8em;
`;

const Connector = styled(PathEnd)`
  margin-top: 0.75em;
`;

export class MaxValue extends React.Component {
  PreviousAnswer = () => (
    <PreviousAnswerContentPicker
      answerId={this.props.answerId}
      onSubmit={this.handlePreviousAnswerChange}
      selectedContentDisplayName={get(
        this.props.maxValue.previousAnswer,
        "displayName"
      )}
      path="answer.validation.maxValue.availablePreviousAnswers"
    />
  );

  Custom = () => (
    <ValidationInput
      data-test="max-value-input"
      list="defaultNumbers"
      value={this.props.maxValue.custom}
      type="number"
      onChange={this.handleCustomValueChange}
      max={this.props.limit}
      min={0 - this.props.limit}
    />
  );

  handlePreviousAnswerChange = ({ value: { id } }) => {
    const updateValidationRuleInput = {
      id: this.props.maxValue.id,
      maxValueInput: {
        inclusive: this.props.maxValue.inclusive,
        previousAnswer: id
      }
    };
    this.props.onUpdateAnswerValidation(updateValidationRuleInput);
  };

  handleCustomValueChange = ({ value }) => {
    // clamp value of input to +/- limit
    if (
      value !== "" &&
      !inRange(parseInt(value, 10), 0 - this.props.limit, this.props.limit + 1)
    ) {
      return false;
    }

    const intValue = parseInt(value, 10);

    const updateValidationRuleInput = {
      id: this.props.maxValue.id,
      maxValueInput: {
        inclusive: this.props.maxValue.inclusive,
        custom: isNaN(intValue) ? null : intValue
      }
    };

    this.props.onUpdateAnswerValidation(updateValidationRuleInput);
  };

  handleEntityTypeChange = value => {
    const updateValidationRuleInput = {
      id: this.props.maxValue.id,
      maxValueInput: {
        inclusive: this.props.maxValue.inclusive,
        entityType: value,
        previousAnswer: null,
        custom: null
      }
    };

    this.props.onUpdateAnswerValidation(updateValidationRuleInput);
  };

  handleToggleChange = ({ value }) => {
    const toggleValidationRuleInput = {
      id: this.props.maxValue.id,
      enabled: value
    };

    this.props.onToggleValidationRule(toggleValidationRuleInput);
  };

  handleIncludeChange = ({ value }) => {
    const updateValidationRuleInput = {
      id: this.props.maxValue.id,
      maxValueInput: {
        custom: this.props.maxValue.custom,
        inclusive: value
      }
    };
    this.props.onUpdateAnswerValidation(updateValidationRuleInput);
  };

  renderDisabled = () => <DisabledMessage name="Max value" />;

  renderContent = () => {
    return (
      <Grid>
        <Column cols={3}>
          <ValidationTitle>Max Value is</ValidationTitle>
          <Connector />
        </Column>
        <Column cols={8}>
          <ValidationPills
            entityType={this.props.maxValue.entityType}
            onEntityTypeChange={this.handleEntityTypeChange}
            PreviousAnswer={this.PreviousAnswer}
            Custom={this.Custom}
          />
          <InlineField>
            <ToggleSwitch
              id="max-value-include"
              name="max-value-include"
              onChange={this.handleIncludeChange}
              checked={this.props.maxValue.inclusive}
            />
            <Label inline htmlFor="max-value-include">
              Include this number
            </Label>
          </InlineField>
        </Column>
      </Grid>
    );
  };

  render() {
    const { maxValue } = this.props;

    return (
      <ValidationView
        onToggleChange={this.handleToggleChange}
        enabled={maxValue.enabled}
        data-test="max-value-view"
      >
        {maxValue.enabled ? this.renderContent() : this.renderDisabled()}
      </ValidationView>
    );
  }
}

MaxValue.defaultProps = {
  limit: 999999999
};

MaxValue.propTypes = {
  answerId: PropTypes.string.isRequired,
  maxValue: propType(MaxValueValidationRule).isRequired,
  answerType: PropTypes.oneOf(Object.values(answerTypes)).isRequired,
  onUpdateAnswerValidation: PropTypes.func.isRequired,
  onToggleValidationRule: PropTypes.func.isRequired,
  limit: PropTypes.number
};

const withQuestionPageEditing = flowRight(
  withApollo,
  withUpdateAnswerValidation,
  withToggleAnswerValidation
);

export const MaxValueWithAnswer = props => (
  <ValidationContext.Consumer>
    {({ answer }) => (
      <MaxValue
        answerId={answer.id}
        maxValue={answer.validation.maxValue}
        answerType={answer.type}
        {...props}
      />
    )}
  </ValidationContext.Consumer>
);

export default withQuestionPageEditing(MaxValueWithAnswer);
