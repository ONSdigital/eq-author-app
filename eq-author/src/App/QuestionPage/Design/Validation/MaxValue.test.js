import React from "react";
import { shallow } from "enzyme";

import { MaxValue } from "App/QuestionPage/Design/Validation/MaxValue";
import { ValidationPills } from "App/QuestionPage/Design/Validation/ValidationPills";

import { CUSTOM, PREVIOUS_ANSWER } from "constants/validation-entity-types";
import { NUMBER } from "constants/answer-types";

const createWrapper = (props, render = shallow) =>
  render(<MaxValue {...props} />);

describe("MaxValue", () => {
  let props, wrapper, onUpdateAnswerValidation, onToggleValidationRule;
  beforeEach(() => {
    onUpdateAnswerValidation = jest.fn();
    onToggleValidationRule = jest.fn();

    props = {
      maxValue: {
        id: "1",
        enabled: true,
        custom: 123,
        inclusive: true,
        entityType: CUSTOM,
        previousAnswer: null
      },
      onUpdateAnswerValidation: onUpdateAnswerValidation,
      onToggleValidationRule: onToggleValidationRule,
      limit: 999999999,
      answerType: NUMBER,
      answerId: "1"
    };

    wrapper = createWrapper(props);
  });

  it("should render with content", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render with disabled content", () => {
    props.maxValue.enabled = false;
    expect(createWrapper(props)).toMatchSnapshot();
  });

  it("should correctly handle toggle change", () => {
    wrapper.simulate("toggleChange", { value: false });

    expect(onToggleValidationRule).toHaveBeenCalledWith({
      id: props.maxValue.id,
      enabled: false
    });
  });

  it("should correctly handle include change", () => {
    wrapper.find("#max-value-include").simulate("change", { value: false });

    expect(onUpdateAnswerValidation).toHaveBeenCalledWith({
      id: props.maxValue.id,
      maxValueInput: {
        custom: props.maxValue.custom,
        inclusive: false
      }
    });
  });

  it("should correctly handle entity type change", () => {
    wrapper.find(ValidationPills).simulate("entityTypeChange", PREVIOUS_ANSWER);

    expect(onUpdateAnswerValidation).toHaveBeenCalledWith({
      id: props.maxValue.id,
      maxValueInput: {
        inclusive: props.maxValue.inclusive,
        entityType: PREVIOUS_ANSWER,
        previousAnswer: null,
        custom: null
      }
    });
  });

  it("should correctly handle previous answer", () => {
    const previousAnswer = {
      id: 1
    };
    const PreviousAnswer = wrapper.find(ValidationPills).prop("PreviousAnswer");
    shallow(<PreviousAnswer />).simulate("submit", {
      name: "previousAnswer",
      value: previousAnswer
    });

    expect(onUpdateAnswerValidation).toHaveBeenCalledWith({
      id: props.maxValue.id,
      maxValueInput: {
        inclusive: props.maxValue.inclusive,
        previousAnswer: previousAnswer.id
      }
    });
  });

  describe("Custom Input", () => {
    let CustomInput, customInputWrapper;

    beforeEach(() => {
      CustomInput = wrapper.find(ValidationPills).prop("Custom");
      customInputWrapper = shallow(<CustomInput />);
    });

    it("should correctly handle max value changes with in range values", () => {
      customInputWrapper.simulate("change", {
        value: 1
      });

      expect(onUpdateAnswerValidation).toHaveBeenCalledWith({
        id: props.maxValue.id,
        maxValueInput: {
          inclusive: props.maxValue.inclusive,
          custom: 1
        }
      });
    });

    it("should correctly handle max value change with empty string", () => {
      customInputWrapper.simulate("change", { value: "" });

      expect(onUpdateAnswerValidation).toHaveBeenCalledWith({
        id: props.maxValue.id,
        maxValueInput: {
          inclusive: props.maxValue.inclusive,
          custom: null
        }
      });
    });

    it("should correctly coerce string inputs to integers", () => {
      customInputWrapper.simulate("change", { value: "1" });

      expect(onUpdateAnswerValidation).toHaveBeenCalledWith({
        id: props.maxValue.id,
        maxValueInput: {
          inclusive: props.maxValue.inclusive,
          custom: 1
        }
      });
    });

    it("should correctly handle max value change with out of range values", () => {
      customInputWrapper.simulate("change", { value: 1000000000 });
      expect(onUpdateAnswerValidation).not.toHaveBeenCalled();

      customInputWrapper.simulate("change", { value: -1000000000 });
      expect(onUpdateAnswerValidation).not.toHaveBeenCalled();

      customInputWrapper.simulate("change", { value: 999999999 });
      expect(onUpdateAnswerValidation).toHaveBeenCalled();

      customInputWrapper.simulate("change", { value: -999999999 });
      expect(onUpdateAnswerValidation).toHaveBeenCalled();
    });
  });
});
