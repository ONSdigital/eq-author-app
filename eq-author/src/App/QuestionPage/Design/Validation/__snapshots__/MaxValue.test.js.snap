// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MaxValue should render with content 1`] = `
<ValidationView
  data-test="max-value-view"
  enabled={true}
  onToggleChange={[Function]}
>
  <Grid
    align="top"
    fillHeight={true}
  >
    <Column
      cols={3}
    >
      <ValidationTitle>
        Max Value is
      </ValidationTitle>
      <MaxValue__Connector />
    </Column>
    <Column
      cols={8}
    >
      <ValidationPills
        Custom={[Function]}
        PreviousAnswer={[Function]}
        entityType="Custom"
        onEntityTypeChange={[Function]}
      />
      <MaxValue__InlineField>
        <ToggleSwitch
          checked={true}
          id="max-value-include"
          name="max-value-include"
          onChange={[Function]}
        />
        <Label
          bold={true}
          htmlFor="max-value-include"
          inline={true}
        >
          Include this number
        </Label>
      </MaxValue__InlineField>
    </Column>
  </Grid>
</ValidationView>
`;

exports[`MaxValue should render with disabled content 1`] = `
<ValidationView
  data-test="max-value-view"
  enabled={false}
  onToggleChange={[Function]}
>
  <DisabledMessage
    name="Max value"
  />
</ValidationView>
`;
